[CBRD-25975] Add a sql test case to covering incorrect hints

LyhengCHHAY opened the pull request:
Refer : http://jira.cubrid.org/browse/CBRD-25975

Issue : `Core dump occurs when entering backtick in SQL hint` 
Tested Ver. `CUBRID 11.5 (11.5.0.1893-61bef33) (64bit release build for Linux)`

All scenarios in this TC will be handled by ignoring incorrect hints without a system crash or core dump. 
Scenarios Include: 

1. Valid hint
2. Nested hint
3. Invalid but safe hint
4. Malformed hint
5. Escaped backtick
6. Empty hint
7. Random content
8. Special characters
9. Foreign language
10. Using other hints rather than use_hash such as (use_nl, use_merge, no_use_hash, no_merge)
11. Malformed hint in join
12. Unclosed hint

junsklee commented:
Please change this PR to draft before I approve it.

junsklee commented:
The scenario description comments should be an evaluate statement. Also, there is no need to add in the "Scenario" portion, as it adds unnecessary clutter.

E.g.,

as-is:
```
-- Scenario 1: Valid hint
```
to-be:
```
evalulate '1 - Valid Hint';
```

junsklee commented:
You should show the query plan your queries in this test case, as it is essential to know if the hints were applied or not. Refer to the sql_guide on how to execute this.

junsklee commented on sql/_13_issues/_25_2h/cases/cbrd_25975.sql:
```diff
@@ -0,0 +1,66 @@
+
+-- Scenario 2: Nested hint
+SELECT /*+ recompile use_hash(testtbl, testtbl) */ * FROM testtbl;
+-- Expected: Parsed correctly if supported; else ignored
```
Move this line to the top of the query. You may include it in the evaluate statement if you believe it makes your answer more legible.

junsklee commented on sql/_13_issues/_25_2h/cases/cbrd_25975.sql:
```diff
@@ -0,0 +1,66 @@
+
+-- Scenario 3: Invalid but safe hint
+SELECT /*+ recompile unknown_hint */ * FROM testtbl;
+-- Expected: Query runs; hint ignored; 
```
Move this line to the top of the query. You may include it in the evaluate statement if you believe it makes your answer more legible.

junsklee commented on sql/_13_issues/_25_2h/cases/cbrd_25975.sql:
```diff
@@ -0,0 +1,66 @@
+SELECT /*+ recompile use_hash( ` ) */ * FROM testtbl;
+SELECT /*+ recompile use_hash( [ ) */ * FROM testtbl;
+SELECT /*+ recompile use_hash( " ) */ * FROM testtbl;
+-- Expected: Syntax error or ignored; no crash
```
Move this line to the top of the query. You may include it in the evaluate statement if you believe it makes your answer more legible.

junsklee commented on sql/_13_issues/_25_2h/cases/cbrd_25975.sql:
```diff
@@ -0,0 +1,66 @@
+
+-- Scenario 5: Escaped backtick
+SELECT /*+ recompile use_hash(\`) */ * FROM testtbl;
+-- Expected: Syntax error or ignored; no crash
```
Move this line to the top of the query. You may include it in the evaluate statement if you believe it makes your answer more legible.

junsklee commented on sql/_13_issues/_25_2h/cases/cbrd_25975.sql:
```diff
@@ -0,0 +1,66 @@
+
+-- Scenario 6: Empty hint
+SELECT /*+ recompile */ * FROM testtbl;
+-- Expected: Query runs normally; no effect
```
Move this line to the top of the query. You may include it in the evaluate statement if you believe it makes your answer more legible.

junsklee commented on sql/_13_issues/_25_2h/cases/cbrd_25975.sql:
```diff
@@ -0,0 +1,66 @@
+
+-- Scenario 7: Random content
+SELECT /*+ recompile 123abc */ * FROM testtbl;
+-- Expected: Query runs; hint ignored; 
```
Move this line to the top of the query. You may include it in the evaluate statement if you believe it makes your answer more legible.

junsklee commented on sql/_13_issues/_25_2h/cases/cbrd_25975.sql:
```diff
@@ -0,0 +1,66 @@
+
+-- Scenario 8: Special characters
+SELECT /*+ recompile use_hash(@!$%) */ * FROM testtbl;
+-- Expected: Syntax error or ignored; no crash
```
Move this line to the top of the query. You may include it in the evaluate statement if you believe it makes your answer more legible.

junsklee commented on sql/_13_issues/_25_2h/cases/cbrd_25975.sql:
```diff
@@ -0,0 +1,66 @@
+
+-- Scenario 9 : Foreign language 
+SELECT /*+ recompile use_hash (កខគឃង) */ * FROM testtbl;
+-- Expected : Query runs normally; no effect
```
Move this line to the top of the query. You may include it in the evaluate statement if you believe it makes your answer more legible.

junsklee commented on sql/_13_issues/_25_2h/cases/cbrd_25975.sql:
```diff
@@ -0,0 +1,66 @@
+SELECT /*+ recompile use_merge( ` ) */ * FROM testtbl;
+SELECT /*+ recompile no_use_hash( [ ) */ * FROM testtbl;
+SELECT /*+ recompile no_merge( " ) */ * FROM testtbl;
+-- Expected: Syntax error or ignored; no crash
```
Move this line to the top of the query. You may include it in the evaluate statement if you believe it makes your answer more legible.

junsklee commented on sql/_13_issues/_25_2h/cases/cbrd_25975.sql:
```diff
@@ -0,0 +1,66 @@
+
+-- Malformed hint in join
+SELECT /*+ recompile use_hash(`) */ * FROM t1, t2 WHERE t1.c1 = t2.c1;
+-- Expected: Syntax error or ignored; no crash
```
Move this line to the top of the query. You may include it in the evaluate statement if you believe it makes your answer more legible.

junsklee commented on sql/_13_issues/_25_2h/cases/cbrd_25975.sql:
```diff
@@ -0,0 +1,66 @@
+
+-- Unclosed hint
+SELECT /*+ recompile use_hash(t1, t2 */ * FROM t1, t2 WHERE t1.c1 = t2.c1;
+-- Expected: Syntax error or ignored; no crash
```
Move this line to the top of the query. You may include it in the evaluate statement if you believe it makes your answer more legible.

junsklee commented:
You should read and verify the query plans and confirm that they return the expected results, otherwise you should either comment on the issue, , ask a peer QA, or create a separate issue.

junsklee commented on sql/_13_issues/_25_2h/cases/cbrd_25975.sql:
```diff
@@ -0,0 +1,66 @@
+-- Expected : Query runs normally; no effect
+
+-- Using other hints rather than use_hash
+SELECT /*+ recompile use_nl( ) */ * FROM testtbl;
```
you should be using backticks (`) for the other hints, as per the issue's premise.

junsklee commented on sql/_13_issues/_25_2h/cases/cbrd_25975.sql:
```diff
@@ -0,0 +1,66 @@
+-- Using other hints rather than use_hash
+SELECT /*+ recompile use_nl( ) */ * FROM testtbl;
+SELECT /*+ recompile use_merge( ` ) */ * FROM testtbl;
+SELECT /*+ recompile no_use_hash( [ ) */ * FROM testtbl;
```
you should be using backticks (`) for the no_use_hash hint, as per the issue's premise.

junsklee commented on sql/_13_issues/_25_2h/cases/cbrd_25975.sql:
```diff
@@ -0,0 +1,66 @@
+SELECT /*+ recompile use_nl( ) */ * FROM testtbl;
+SELECT /*+ recompile use_merge( ` ) */ * FROM testtbl;
+SELECT /*+ recompile no_use_hash( [ ) */ * FROM testtbl;
+SELECT /*+ recompile no_merge( " ) */ * FROM testtbl;
```
you should be using backticks (`) for the no_merge hint, as per the issue's premise.

junsklee commented:
I'd suggest adding in test cases that contain two or more hints with a backtick.
Ex)
- 2 hints with a backtick on only one of the hint
- 2 hints with a backtick on both hints
- 2 different hints (e.g., use_hash and no_merge) with a backtick on each hint
- 3 hints with a backtick

junsklee commented:
You forgot to drop your tables at the end.

```
DROP TABLE t1;
DROP TABLE t2;
DROP TABLE  testtbl;
```

junsklee commented:
Suggested table and column name improvements:


- Although there are no strict naming conventions to tables, please keep them consistent (testtbl is not consistent with the formatting of tables t1 and t2). 

- Also, DO NOT include numbers in your table and column names, especially when you're viewing query plans. CTP converts digits into question marks in the query plans of answer files (e.g., t1 => t?, t2 => t?, and c1 => c?), making it extremely difficult to distinguish which table is which.

LyhengCHHAY commented:
@junsklee 
Review reflected: 

1. Moving query order
2. Using backtick in other hints as for the issue's premise
3. Added four new scenarios 
4. Drop all tables  at the end of the test
5. Changing table and column name tableb1 -> tableA, t1 -> tableB etc.

One haven't reflected

1. Query plan - I follow the sql doc by creating test-case-name.queryPlan but there is nothing print to the file. I will up more tomorrow but if you know the cause of that please let me know I would appreciate it.

junsklee commented:
> One haven't reflected
> 
> 1. Query plan - I follow the sql doc by creating test-case-name.queryPlan but there is nothing print to the file. I will up more tomorrow but if you know the cause of that please let me know I would appreciate it.

It looks like the query plan was indeed printed in your answer file. 
```
==================================================

 1 - Nested hint     

 ===================================================
 columna    
 1     

 Query plan:
 sscan
     class: tableA node[?]
     cost:  ? card ?


 Query stmt:
 select /*+ USE_HASH(tableA, tableA) */ tableA.columna from tableA tableA
 ===================================================
```

Please add and commit your empty _sql/_13_issues/_25_2h/cases/cbrd_25975.queryplan_ file to the PR;  it has yet been uploaded.

junsklee commented:
Please use snake case instead of camel case for column names. It's difficult to read the query plan because SQL results and query plans often ignore casings. For instance, in the query plan output, columnA becomes columna, columnB becomes columnb, etc. which makes it almost illegible, rendering it unnecessarily difficult to analyze).

Suggested Change:
`tableA -> tbl_a`
`columnA -> col_a`
`columnB -> col_b`
...

junsklee commented:
The expected comment (e.g., _-- Expected: Syntax error or ignored; no crash_) review has not been reflected. Each expected comment should be above the related SELECT query.

Ex)
as-is
```
evaluate '2 - Invalid but safe hint';
 SELECT /*+ recompile unknown_hint */ * FROM tableA;
 -- Expected: Query runs; hint ignored; 
```
 
 to-be
```
 evaluate '2 - Invalid but safe hint';
 -- Expected: Query runs; hint ignored; 
 SELECT /*+ recompile unknown_hint */ * FROM tableA;
```

junsklee commented:
Also, your Evaluate statement could be improved by including the hints used. It may even include the entire query itself (as for your scenarios, it is short enough to include in the evaluate statement as they're mostly one line).

ex)
```
 evaluate '7 - Special characters: /*+ recompile use_hash(@!$%) */';
```

OR
```
 evaluate '7 - Special characters: [SELECT /*+ recompile use_hash(@!$%) */ * FROM tbl_a;]';
```

The ultimate goal should be to be able to understand your intent just by looking at the answer file.

LyhengCHHAY commented:
@junsklee 
Updated Include: 

1. Added cbrd_25975.queryplan file 
2. Used snake case for table and column name
3. Moved 'Expected' statement to under 'evaluate' statement 
4. Improved evaluate statement for hints use in the answer file

kwonhoil commented on sql/_13_issues/_25_2h/cases/cbrd_25975.sql:
```diff
@@ -0,0 +1,104 @@
+-- Table Setup
```
"cbrd_25975.queryplan" 파일이 필요한지 확인해 주세요.
불필요하다면 삭제해 주세요.

junsklee replied:
힌트가 제대로 적용이 됐는지 확인하는 이슈여서 쿼리플랜 출력이 필요할 것 같습니다. 최근 커밋에 .answer 파일에 쿼리플랜이 출력이 안되고있어서 PR 담당자에게 말씀드리겠습니다.

junsklee commented:
It looks like you've added the empty queryPlan file but the query plan output is not showing up in the result queries in the answer file. 

I believe it has to do with your queryPlan file naming (incorrect casing). 

as-is:
`cbrd_25975.queryplan`

to-be:
`cbrd_25975.queryPlan`

Please investigate what query plans are, and ensure that your answer files are correct by matching them up with your queries, and comparing it against your previous commit. When the current commit is compared against the previous commit (see: https://github.com/CUBRID/cubrid-testcases/pull/2269/commits/cb2bd17fef040f664051fe7e79a810118db3fb5a), I can see that the query plan details have been removed, which should've been identified as an anomaly that needed to be fixed before publishing your commit.

LyhengCHHAY commented:
@junsklee 
Update include: 

1. Corrected query file name (as-is cbrd_25975.queryplan - to-be cbrd_25975.querypPlan )
2. Printing query plan context inside answer file

junsklee commented:
A comment has been added to the issue regarding additional test cases. Please review and add them to your TC: http://jira.cubrid.org/browse/CBRD-25975?focusedCommentId=4772036&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-4772036

LyhengCHHAY commented:
> A comment has been added to the issue regarding additional test cases. Please review and add them to your TC: http://jira.cubrid.org/browse/CBRD-25975?focusedCommentId=4772036&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-4772036

added

junsklee commented on sql/_13_issues/_25_2h/cases/cbrd_25975.sql:
```diff
@@ -99,6 +99,125 @@ evaluate '16 - Valid hint: SELECT /*+ recompile use_hash(tbl_a) */ * FROM tbl_a;
 
+evaluate '17 - Empty parentheses: [SELECT /*+ recompile use_hash() */ * FROM tbl_b JOIN tbl_c ON tbl_b.col_b=tbl_c.col_b;]';
+-- Expected: Syntax error or ignored
+SELECT /*+ recompile use_hash() */ * FROM tbl_b JOIN tbl_c ON tbl_b.col_b=tbl_c.col_b; FROM t1 JOIN t2 ON t1.col_d=t2.col_d; 
```
`FROM t1 JOIN t2 ON t1.col_d=t2.col_d;`

I think this is an extra line that you didn't intend to include. Please remove this.

junsklee commented:
`-- Expected: Syntax error or ignored`
Your comment above describe two completely different behaviors. I suggest you should mention either "Syntax error" or "Hint ignored", not both.

LyhengCHHAY commented:
> `-- Expected: Syntax error or ignored` Your comment above describe two completely different behaviors. I suggest you should mention either "Syntax error" or "Hint ignored", not both.

`-- Expected: ...` comment had been modified
